modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ValueProviderModelValidator.java

```diff
@@ -146,4 +196,46 @@ private void validateOptionsResolver(ParameterModel param, ValueProviderFactoryM
       }
     }
   }
+
+  private void addValueProvider(String providerName, ValueProviderModel valueProviderModel,
+                                Map<String, ValueProviderInformation> valueProviders, ProblemsReporter problemsReporter,
+                                ParameterizedModel model, ParameterModel parameterModel) {
+    if (valueProviders.containsKey(providerName)) {
+      String storedProviderId = valueProviders.get(providerName).getValueProviderModel().getProviderId();
+      String providerId = valueProviderModel.getProviderId();
+      if (!providerId.equals(storedProviderId)) {
```

Comment:
if this method is called add, then it should just add, not validate. It's also weird that  you also have the  `validateProviderIds ` method traversing the map again validating other aspects. You should either completely separate data gathering from data validation all-together, or validate as you gather, but this mix up is not a good idea and ultimately needs to the algorithm being highly inefficient.

Reply:
What I am validating here is that no implementation of value provider is associated with more than one id. This should not happen unless the value provider implementation uses a random component for the getId() method. With the structure that I am storing the values, this information is not available later on. I will make this information available to validate it later so that gathering and validation are completely separated.